### PR TITLE
refactor: Fadec rewrite using Cpp framework

### PR DESCRIPTION
**Pull request by frankkopp**
## Summary of Changes
Migration of the current fadec code for the A32NX and A380X to the cpp framework with cleanup and additional commenting for better understanding and maintainability.  

No change in logic and also not further modularized into shared and specific parts as most of the A380X fadec code still uses the A32NX polynomials and tables. As this will surely change any further sharing of code should be done after when it is clear which parts of the code are really sharable. 

Performance has improved a little as reading and writing of data from/to the sim has been optimized. 

Discord username (if different from GitHub): cdr_maverick

## Testing instructions
Full regression tests on the A32NX with full flights from c&d, runway and the air. 
Also fuel handling (fueling, saving and loading of fuel per ATC ID) needs to be tested. 

Same for the A380X but to a lesser extend as the fadec code is not unfinished yet anyway. 
As on the fuel page in the A380X the order of the tanks changed please also test and look at this.  

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: fbw-common/src/systems/instruments/src/EFB/Ground/Pages/Fuel/A380_842/A380Fuel.tsx
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Ground/Pages/Fuel/A380_842/A380Fuel.tsx:1]: Should this diff be in this PR? This seems like a EFB related change not a CPP framework related. Unless this is intended as part of this PR?

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+   * @return The calculated N3 value in percent.
+   */
+  static double startN3(double currentSimN3, double previousN3, double idleN3) {
+    // Normalize the current N3 percentage by scaling it with the idle N2
+    // percentage and a constant factor.
+    // The constant factor 60.0 is likely derived from empirical data or a mathematical model of the
+    // engine's behavior.
+    double normalizedN3 = currentSimN3 * 60.0 / idleN3;
+
+    // Coefficients for the polynomial used to calculate the N2 percentage.
```
Comment (2hwk) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp:35]:
```suggestion
    // Coefficients for the polynomial used to calculate the N3 percentage.
```

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+ *
+ * TODO: Many of the values/polynomials used in these methods are identical to the A32NX values and
+ *       likely need to be adjusted to match the A380X engine model.
+ */
+class Polynomial_A380X {
+ public:
+  /**
+   * @brief Calculates the N3 value during engine startup using real-life modeled polynomials.
+   *
+   * @param currentSimN3 The current N3 value in percent (taken from the sim's N2 value).
```
Comment (2hwk) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp:23]:
```suggestion
   * @param currentSimN3 The current N3 value in percent (taken from the sim's N3 value).
```

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+   * @param previousN3 The previous N3 value in percent.
+   * @param idleN3 The idle N3 value in percent.
+   * @return The calculated N3 value in percent.
+   */
+  static double startN3(double currentSimN3, double previousN3, double idleN3) {
+    // Normalize the current N3 percentage by scaling it with the idle N2
```
Comment (2hwk) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp:29]:
```suggestion
    // Normalize the current N3 percentage by scaling it with the idle N3
```

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+   */
+  static double oilGulpPct(double thrust) {
+    const double oilGulpCoefficients[3] = {20.1968848, -1.2270302e-4, 1.78442e-8};
+    const double oilGulpPercentage =
+        oilGulpCoefficients[0] + (oilGulpCoefficients[1] * thrust) + (oilGulpCoefficients[2] * (std::pow)(thrust, 2));
+    return oilGulpPercentage / 100;
+  }
+
+  /**
+   * @brief Calculates the Oil Pressure (PSI) based on simulated N2 value.
```
Comment (2hwk) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/Polynomials_A380X.hpp:437]:
```suggestion
   * @brief Calculates the Oil Pressure (PSI) based on simulated N3 value.
```

File: fbw-a380x/src/wasm/fadec_a380x/src/Fadec/FadecSimData_A380X.hpp [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+
+    oilPsiDataPtr[E1] = dm->make_datadefinition_var<OilPsiData>("GENERAL ENG OIL PRESSURE 1", oilPsiE1DataDef, NO_AUTO_UPDATE);
+    oilPsiDataPtr[E2] = dm->make_datadefinition_var<OilPsiData>("GENERAL ENG OIL PRESSURE 2", oilPsiE2DataDef, NO_AUTO_UPDATE);
+    oilPsiDataPtr[E3] = dm->make_datadefinition_var<OilPsiData>("GENERAL ENG OIL PRESSURE 3", oilPsiE3DataDef, NO_AUTO_UPDATE);
+    oilPsiDataPtr[E4] = dm->make_datadefinition_var<OilPsiData>("GENERAL ENG OIL PRESSURE 4", oilPsiE4DataDef, NO_AUTO_UPDATE);
+
+    engineCorrectedN3DataPtr[E1] = dm->make_datadefinition_var<CorrectedN3Data>("TURB ENG CN2 1", engine1CN3DataDef, NO_AUTO_UPDATE);
+    engineCorrectedN3DataPtr[E2] = dm->make_datadefinition_var<CorrectedN3Data>("TURB ENG CN2 2", engine2CN3DataDef, NO_AUTO_UPDATE);
+    engineCorrectedN3DataPtr[E3] = dm->make_datadefinition_var<CorrectedN3Data>("TURB ENG CN2 3", engine3CN3DataDef, NO_AUTO_UPDATE);
+    engineCorrectedN3DataPtr[E4] = dm->make_datadefinition_var<CorrectedN3Data>("TURB ENG CN2 4", engine4CN3DataDef, NO_AUTO_UPDATE);
```
Comment (2hwk) [inline fbw-a380x/src/wasm/fadec_a380x/src/Fadec/FadecSimData_A380X.hpp:270]: Even though this is already covered by a comment to the struct above

 > Note: the sim does not have a direct N3 value, so we use N2 as a proxy

I still think there should be a comment/note above here which again notes and refers to this, as it is important.

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: fbw-common/src/systems/instruments/src/EFB/Ground/Pages/Fuel/A380_842/A380Fuel.tsx
Comment (2hwk) [reply fbw-common/src/systems/instruments/src/EFB/Ground/Pages/Fuel/A380_842/A380Fuel.tsx:1]: As this is intentional, please remember to add a basic testing instruction to check this page's functionality

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: fbw-common/src/systems/instruments/src/EFB/Ground/Pages/Fuel/A380_842/A380Fuel.tsx
Comment (frankkopp) [reply fbw-common/src/systems/instruments/src/EFB/Ground/Pages/Fuel/A380_842/A380Fuel.tsx:1]: will do

**Review by Maximilian-Reuter — APPROVED** [review]

**Review by 2hwk — APPROVED** [review]

**Comment by 2hwk** [comment]
Fix merge conflicts, then reset Ready to Test tag

**Comment by noonebling** [comment]
Quality Assurance Trainee Report

Discord          : noonebling
Object of testing: 8547
Tier of Testing  : 2
Date             : 10/04/2024

Testing Process: 
Full regression tests on the A32NX with full flights from c&d, runway and the air.
Fuel handling (fueling, saving and loading of fuel per ATC ID) needs to be tested.

Negatives:
N/A

Testing Results: 
Passed

Conclusions: 
Works as expected

Media:
N/A

**Pull request merged**